File: src/main/java/org/gradle/plugins/site/SitePlugin.java
package org.gradle.plugins.site;

import org.gradle.api.Action;
import org.gradle.api.Plugin;
import org.gradle.api.Project;
import org.gradle.api.Task;
import org.gradle.api.plugins.JavaBasePlugin;
import org.gradle.api.plugins.JavaPlugin;
import org.gradle.api.plugins.JavaPluginConvention;
import org.gradle.plugins.site.data.EnvironmentDescriptor;
import org.gradle.plugins.site.data.JavaProjectDescriptor;
import org.gradle.plugins.site.data.ProjectDescriptor;
import org.gradle.plugins.site.data.TaskDescriptor;
import org.gradle.plugins.site.tasks.SiteGenerate;

import java.util.concurrent.Callable;

/**
 * A plugin for generating a web page containing derived project information.
 * <p>
 * The default directory for generated web page is {@code $buildDir/docs/site}.
 */
public class SitePlugin implements Plugin<Project> {

    /**
     * The name of the extension for configuring the runtime behavior of the plugin.
     *
     * @see org.gradle.plugins.site.SitePluginExtension
     */
    public static final String EXTENSION_NAME = "site";

    /**
     * The name of task generating the web page containing derived project information.
     *
     * @see org.gradle.plugins.site.tasks.SiteGenerate
     */
    public static final String GENERATE_SITE_TASK_NAME = "generateSite";

    @Override
    public void apply(final Project project) {
        SitePluginExtension sitePluginExtension = project.getExtensions().create(EXTENSION_NAME, SitePluginExtension.class, project);
        sitePluginExtension.getOutputDir().set(project.getLayout().getBuildDirectory().dir("docs/site"));
        SiteGenerate siteGenerateTask = createSiteTask(project, sitePluginExtension);
        siteGenerateTask.getProjectDescriptor().set(project.provider(new Callable<ProjectDescriptor>() {
            @Override
            public ProjectDescriptor call() throws Exception {
                return deriveProjectDescription(project);
            }
        }));
    }

    private ProjectDescriptor deriveProjectDescription(Project project) {
        ProjectDescriptor projectDescriptor = new ProjectDescriptor(project.getName(), project.getGroup().toString(), project.getDescription(), project.getVersion().toString(), new EnvironmentDescriptor(project.getGradle().getGradleVersion()));
        addPluginDescription(project, projectDescriptor);
        addTasksDescription(project, projectDescriptor);
        addJavaDescription(project, projectDescriptor);
        return projectDescriptor;
    }

    private void addJavaDescription(final Project project, final ProjectDescriptor projectDescriptor) {
        project.getPlugins().withType(JavaPlugin.class, new Action<JavaPlugin>() {
            @Override
            public void execute(JavaPlugin javaPlugin) {
                JavaPluginConvention javaConvention = project.getConvention().getPlugin(JavaPluginConvention.class);
                projectDescriptor.setJavaProject(new JavaProjectDescriptor(javaConvention.getSourceCompatibility().toString(), javaConvention.getTargetCompatibility().toString()));
            }
        });
    }

    private void addPluginDescription(Project project, final ProjectDescriptor projectDescriptor) {
        project.getPlugins().all(new Action<Plugin>() {
            @Override
            public void execute(Plugin plugin) {
                projectDescriptor.addPluginClass(plugin.getClass());
            }
        });
    }

    private void addTasksDescription(Project project, final ProjectDescriptor projectDescriptor) {
        project.getTasks().all(new Action<Task>() {
            @Override
            public void execute(Task task) {
                if (task.getGroup() != null) {
                    projectDescriptor.addTask(new TaskDescriptor(task.getName(), task.getPath(), task.getGroup(), task.getDescription()));
                }
            }
        });
    }

    private SiteGenerate createSiteTask(Project project, SitePluginExtension sitePluginExtension) {
        SiteGenerate generateSiteTask = project.getTasks().create(GENERATE_SITE_TASK_NAME, SiteGenerate.class);
        generateSiteTask.setGroup(JavaBasePlugin.DOCUMENTATION_GROUP);
        generateSiteTask.setDescription("Generates a web page containing information about the project.");
        generateSiteTask.getOutputDir().set(sitePluginExtension.getOutputDir());
        generateSiteTask.getCustomData().setWebsiteUrl(sitePluginExtension.getWebsiteUrl());
        generateSiteTask.getCustomData().setVcsUrl(sitePluginExtension.getVcsUrl());
        return generateSiteTask;
    }
}

File: src/main/kotlin/org/gradle/plugins/site/SitePlugin.kt
package org.gradle.plugins.site

import org.gradle.api.Plugin
import org.gradle.api.Project
import org.gradle.api.plugins.JavaBasePlugin
import org.gradle.api.plugins.JavaPlugin
import org.gradle.api.plugins.JavaPluginExtension
import org.gradle.api.tasks.TaskProvider
import org.gradle.plugins.site.data.EnvironmentDescriptor
import org.gradle.plugins.site.data.JavaProjectDescriptor
import org.gradle.plugins.site.data.ProjectDescriptor
import org.gradle.plugins.site.data.TaskDescriptor
import org.gradle.plugins.site.tasks.SiteGenerate

class SitePlugin : Plugin<Project> {
    /**
     * The name of the extension for configuring the runtime behavior of the plugin.
     *
     * @see org.gradle.plugins.site.SitePluginExtension
     */
    val EXTENSION_NAME = "site"

    /**
     * The name of task generating the web page containing derived project information.
     *
     * @see org.gradle.plugins.site.tasks.SiteGenerate
     */
    val GENERATE_SITE_HTML_TASK_NAME = "generateSiteHtml"

    val DEFAULT_OUTPUT_DIR = "docs/site"

    override fun apply(project: Project): Unit = project.run {
        val sitePluginExtension = extensions.create(EXTENSION_NAME, SitePluginExtension::class.java, project)
        sitePluginExtension.outputDir.set(layout.buildDirectory.dir(DEFAULT_OUTPUT_DIR))

        registerSiteTask(sitePluginExtension).configure { siteGenerate ->
            siteGenerate.projectDescriptor.set(provider {
                deriveProjectDescription()
            })
        }
    }

    private fun Project.deriveProjectDescription(): ProjectDescriptor {
        val projectDescriptor = ProjectDescriptor(name, group.toString(), description.orEmpty(), version.toString(), EnvironmentDescriptor(gradle.gradleVersion))
        addPluginDescription(projectDescriptor)
        addTasksDescription(projectDescriptor)
        addJavaDescription(projectDescriptor)
        return projectDescriptor
    }

    private fun Project.addPluginDescription(projectDescriptor: ProjectDescriptor) {
        plugins.all { plugin -> projectDescriptor.addPluginClass(plugin.javaClass) }
    }

    private fun Project.addTasksDescription(projectDescriptor: ProjectDescriptor) {
        tasks.all { task ->
            if (task.group != null) {
                projectDescriptor.addTask(TaskDescriptor(task.name, task.path, task.group!!, task.description ?: ""))
            }
        }
    }

    private fun Project.addJavaDescription(projectDescriptor: ProjectDescriptor) {
        plugins.withType(JavaPlugin::class.java) {
            val java = extensions.getByType(JavaPluginExtension::class.java)
            projectDescriptor.javaProject = JavaProjectDescriptor(java.sourceCompatibility.toString(), java.targetCompatibility.toString())
        }
    }

    private fun Project.registerSiteTask(sitePluginExtension: SitePluginExtension): TaskProvider<SiteGenerate> =
            tasks.register(GENERATE_SITE_HTML_TASK_NAME, SiteGenerate::class.java) { siteGenerate ->
                siteGenerate.group = JavaBasePlugin.DOCUMENTATION_GROUP
                siteGenerate.description = "Generates a web page containing information about the project."
                siteGenerate.outputDir.set(sitePluginExtension.outputDir)
                siteGenerate.customData.setWebsiteUrl(sitePluginExtension.websiteUrl)
                siteGenerate.customData.setVcsUrl(sitePluginExtension.vcsUrl)
            }
}
